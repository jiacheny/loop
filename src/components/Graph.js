import React, {Component} from 'react';

import GroupNode from './UI/GroupNode';
import UserNode from './UI/UserNode';

import {drawGraph} from "./utils/Helper";

class Graph extends Component {

  render() {

    const graphData = drawGraph(this.props.isMemberOfGraph);
    let graph;
    for (const i in graphData) {
      const data = graphData[i];
      if (data.type === "g") {
        graph = [graph, <GroupNode key={data.key} name={this.getName(data.name)} indentation={data.indentation}/>]
      } else if (data.type === "u") {
        graph = [graph, <UserNode key={data.key} name={this.getName(data.name)} indentation={data.indentation}/>];
      }
    }

    return (
      <div id="membershipGraph">

        <div className="">
          <h2>Membership Graph</h2>
        </div>

        <div className="row">
          <div id="graph" className="rounded border border-light col-md-6 mx-auto">
            {graph}
          </div>
        </div>

      </div>
    );
  }

  getName = (key) => {
    return this.props.nodes[key].name;
  }

}

export default Graph;