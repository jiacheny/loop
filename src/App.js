import React, {Component} from 'react';
import './App.css';

import SingleMemberManagement from './components/SingleMemberManagement';
import Graph from './components/Graph';

const headerStyle = {
  height: '100px'
};

class App extends Component {

  constructor() {
    super();
    this.state = {
      nodes: {
        "g1": {name: "Group 1", type: "group"},
        "g2": {name: "Group 2", type: "group"},
        "g3": {name: "Group 3", type: "group"},
        "g4": {name: "Group 4", type: "group"},
        "g5": {name: "Group 5", type: "group"},
        "g6": {name: "Group 6", type: "group"},
        "g7": {name: "Group 7", type: "group"},
        "u1": {name: "User 1", type: "user"},
        "u2": {name: "User 2", type: "user"},
        "u3": {name: "User 3", type: "user"},
        "u4": {name: "User 4", type: "user"},
        "u5": {name: "User 5", type: "user"},
        "u6": {name: "User 6", type: "user"},
        "u7": {name: "User 7", type: "user"}
      },
      isMemberOfGraph: [
        {from: "g2", to: "g1"},
        {from: "g3", to: "g1"},
        {from: "g4", to: "g1"},
        {from: "u1", to: "g2"},
        {from: "u1", to: "g5"},
        {from: "u2", to: "g2"},
        {from: "u3", to: "g3"},
        {from: "u4", to: "g3"},
        {from: "u5", to: "g4"},
        {from: "u6", to: "g4"},
        {from: "u7", to: "g5"}
      ],
      newNodeNameExist: false
    }
  }

  render() {
    return (

      <div className="App">

        <div style={headerStyle} className="bg-primary text-white d-flex justify-content-center align-items-center">
          <h1>Loop Membership Management</h1>
        </div>

        <SingleMemberManagement
          newNodeNameExist={this.state.newNodeNameExist}
          createClicked={this.createClicked.bind(this)}
          deleteClicked={this.deleteClicked.bind(this)}
        />

        <div id="memberAssignment" className="form-row p-3 d-flex justify-content-center">

          <div className="col-md-2">
            <label>Member</label>
            <select className="custom-select">
              <option>Select member</option>
              <option disabled={true}>--- users ---</option>
              <option disabled={true}>--- groups ---</option>
            </select>
          </div>


          <div className="col-md-2">
            <label>Group</label>
            <select className="custom-select">
              <option>Select group</option>
            </select>
          </div>

          <div className="col-md-1 align-self-end">
            <button className="btn btn-primary btn-block">
              Add
            </button>
          </div>

          <div className="col-md-1 align-self-end">
            <button className="btn btn-primary btn-block">
              Remove
            </button>
          </div>
        </div>

        <Graph nodes={this.state.nodes} isMemberOfGraph={this.state.isMemberOfGraph}/>

      </div>

    );
  }

  createClicked(newNode) {

    const nodes = this.state.nodes;

    const doesExist = !(Object.values(nodes).filter(e =>
      e.type === newNode.type && e.name === newNode.name
    ).length === 0);

    if (!doesExist) {
      let key = this.getKeyForNewNode(newNode);
      nodes[key] = newNode;
      this.setState({
        nodes: nodes,
        newNodeNameExist: false
      });
    } else {
      this.setState({
        newNodeNameExist: true
      })
    }

  };

  deleteClicked(node) {
    const nodes = this.state.nodes;
    const key = Object.keys(nodes).find(k =>
      nodes[k].name === node.name && nodes[k].type === node.type
    );
    if (key) {
      delete nodes[key];
    }

  }

  getRandomKey(newNode) {
    let key = "";
    let id = Math.round(Math.random() * 100);
    if (newNode.type === 'group') {
      key = "g" + id;
    } else if (newNode.type === "user") {
      key = "u" + id;
    }
    return key;
  }

  getKeyForNewNode(newNode) {
    let key = this.getRandomKey(newNode);
    while (key in this.state.nodes) {
      key = this.getRandomKey(newNode);
    }
    return key;
  }

}

export default App;
