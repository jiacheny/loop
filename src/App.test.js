import React from 'react';
import ReactDOM from 'react-dom';
import {configure, shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from './App';
import Graph from './components/Graph';
import renderer from 'react-test-renderer';


configure({adapter: new Adapter()});

describe('<App />', () => {

  let app;
  let nameInput, typeSelect, createButton, deleteButton;

  beforeEach(() => {
    app = mount(<App/>);
    nameInput = app.find("input").at(0);
    typeSelect = app.find("select").at(0);
    createButton = app.find("button").at(1);
    deleteButton = app.find("button").at(3);
  });


  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App/>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('adds a new user successfully', () => {

    const newUserName = "Test User";

    nameInput.instance().value = newUserName;
    nameInput.simulate("change");

    typeSelect.instance().value = "user";
    typeSelect.simulate("change");

    createButton.simulate("click");

    expect(Object.values(app.state("nodes")).filter(n => n.name === newUserName && n.type === 'user').length).toEqual(1);

  });

  it('add a new user with name already exists', () => {

    const initialNumberOfUser = Object.values(app.state("nodes")).filter(n => n.type === 'user').length;

    nameInput.instance().value = "User 1";
    nameInput.simulate("change");

    typeSelect.instance().value = "user";
    typeSelect.simulate("change");

    createButton.simulate("click");

    expect(Object.values(app.state("nodes")).filter(n => n.type === 'user').length).toEqual(initialNumberOfUser);
    expect(app.state("newNodeNameExist")).toEqual(true);
    expect(app.contains("This name already exists.")).toEqual(true);
    expect(app.contains("Please select a type.")).toEqual(false);

  });

  it('add a new group successfully', () => {

    const newGroupName = "Test Group";
    nameInput.instance().value = "Test Group";
    nameInput.simulate("change");

    typeSelect.instance().value = "group";
    typeSelect.simulate("change");

    const createButton = app.find("button").at(1);
    createButton.simulate("click");

    expect(Object.values(app.state("nodes")).filter(n => n.name === newGroupName && n.type === 'group').length).toEqual(1);
  });

  it('add a new group with name already exists', () => {

    const initialNumberOfGroup = Object.values(app.state("nodes")).filter(n => n.type === 'group').length;

    nameInput.instance().value = "Group 1";
    nameInput.simulate("change");

    typeSelect.instance().value = "group";
    typeSelect.simulate("change");

    createButton.simulate("click");

    expect(Object.values(app.state("nodes")).filter(n => n.type === 'group').length).toEqual(initialNumberOfGroup);
    expect(app.state("newNodeNameExist")).toEqual(true);
    expect(app.contains("This name already exists.")).toEqual(true);
    expect(app.contains("Please select a type.")).toEqual(false);

  });

  it('click create with no name and no type', () => {
    createButton.simulate("click");
    expect(app.contains("Please enter a name.")).toEqual(true);
    expect(app.contains("Please select a type.")).toEqual(true);
  });

  it('remove an existing user successfully', () => {

    const initialNumberOfUser = Object.values(app.state("nodes")).filter(n => n.type === 'user').length;

    nameInput.instance().value = "User 1";
    nameInput.simulate("change");

    typeSelect.instance().value = "user";
    typeSelect.simulate("change");

    deleteButton.simulate("click");
    expect(Object.values(app.state("nodes")).filter(n => n.type === 'user').length).toEqual(initialNumberOfUser - 1);

  });

  it('remove an existing group successfully', () => {

    const initialNumberOfGroup = Object.values(app.state("nodes")).filter(n => n.type === 'group').length;

    nameInput.instance().value = "Group 1";
    nameInput.simulate("change");

    typeSelect.instance().value = "group";
    typeSelect.simulate("change");

    deleteButton.simulate("click");
    expect(Object.values(app.state("nodes")).filter(n => n.type === 'group').length).toEqual(initialNumberOfGroup - 1);

  });

  it('graph matches the snapshot', () => {
    const mockState = {
      nodes: {
        "g1": {name: "Group 1", type: "group"},
        "g2": {name: "Group 2", type: "group"},
        "g3": {name: "Group 3", type: "group"},
        "g4": {name: "Group 4", type: "group"},
        "g5": {name: "Group 5", type: "group"},
        "g6": {name: "Group 6", type: "group"},
        "g7": {name: "Group 7", type: "group"},
        "u1": {name: "User 1", type: "user"},
        "u2": {name: "User 2", type: "user"},
        "u3": {name: "User 3", type: "user"},
        "u4": {name: "User 4", type: "user"},
        "u5": {name: "User 5", type: "user"},
        "u6": {name: "User 6", type: "user"},
        "u7": {name: "User 7", type: "user"}
      },
      isMemberOfGraph: [
        {from: "g2", to: "g1"},
        {from: "g3", to: "g1"},
        {from: "g4", to: "g1"},
        {from: "u1", to: "g2"},
        {from: "u1", to: "g5"},
        {from: "u2", to: "g2"},
        {from: "u3", to: "g3"},
        {from: "u4", to: "g3"},
        {from: "u5", to: "g4"},
        {from: "u6", to: "g4"},
        {from: "u7", to: "g5"}
      ],
      newNodeNameExist: false
    };
    const appTree = renderer.create(<Graph nodes={mockState.nodes} isMemberOfGraph={mockState.isMemberOfGraph}/>).toJSON();
    expect(appTree).toMatchSnapshot();
  });

});