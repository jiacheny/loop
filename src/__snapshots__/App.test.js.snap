// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> graph matches the snapshot 1`] = `
<div
  id="membershipGraph"
>
  <div
    className=""
  >
    <h2>
      Membership Graph
    </h2>
  </div>
  <div
    className="row"
  >
    <div
      className="rounded border border-light col-md-6 mx-auto"
      id="graph"
    >
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 0,
            }
          }
        />
        <div
          className="bg-primary text-white rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "margin": "5px",
              "width": "150px",
            }
          }
        >
          Group 1
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <div
          className="bg-primary text-white rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "margin": "5px",
              "width": "150px",
            }
          }
        >
          Group 2
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 100,
            }
          }
        />
        <div
          className="border border-primary rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "marginBottom": "5px",
              "width": "150px",
            }
          }
        >
          User 1
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 100,
            }
          }
        />
        <div
          className="border border-primary rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "marginBottom": "5px",
              "width": "150px",
            }
          }
        >
          User 2
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <div
          className="bg-primary text-white rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "margin": "5px",
              "width": "150px",
            }
          }
        >
          Group 3
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 100,
            }
          }
        />
        <div
          className="border border-primary rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "marginBottom": "5px",
              "width": "150px",
            }
          }
        >
          User 3
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 100,
            }
          }
        />
        <div
          className="border border-primary rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "marginBottom": "5px",
              "width": "150px",
            }
          }
        >
          User 4
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <div
          className="bg-primary text-white rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "margin": "5px",
              "width": "150px",
            }
          }
        >
          Group 4
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 100,
            }
          }
        />
        <div
          className="border border-primary rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "marginBottom": "5px",
              "width": "150px",
            }
          }
        >
          User 5
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 100,
            }
          }
        />
        <div
          className="border border-primary rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "marginBottom": "5px",
              "width": "150px",
            }
          }
        >
          User 6
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 0,
            }
          }
        />
        <div
          className="bg-primary text-white rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "margin": "5px",
              "width": "150px",
            }
          }
        >
          Group 5
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <div
          className="border border-primary rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "marginBottom": "5px",
              "width": "150px",
            }
          }
        >
          User 1
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <div
          className="border border-primary rounded d-flex justify-content-center align-items-center"
          style={
            Object {
              "height": "50px",
              "marginBottom": "5px",
              "width": "150px",
            }
          }
        >
          User 7
        </div>
      </div>
    </div>
  </div>
</div>
`;
